### ISSUE TITLE
Large numerical errors in torch.einsum and torch.matmul on A40 GPU

### ISSUE THREAD

**Issue by hchau630** [pytorch/pytorch#78275]
### 🐛 Describe the bug

I ran the following code on a 1080 Ti and an A40 GPUs respectively (this is done on a cluster). The 1080 Ti gives the expected result while the A40 GPU gives me the wrong result.

```
import torch

device = torch.device('cuda')
z = torch.tensor([0.0, 3.14]) 
S = torch.tensor([[1.0, 0.0],
                  [0.0, 1.0]])

print(torch.einsum('i,ij,j->',z,S,z))
print(z.t() @ S @ z)
print(z[0]*S[0,0]*z[0]+z[0]*S[0,1]*z[1]+z[1]*S[1,0]*z[0]+z[1]*S[1,1]*z[1])

z = z.to(device)
S = S.to(device)

print(torch.einsum('i,ij,j->',z,S,z))
print(z.t() @ S @ z)
print(z[0]*S[0,0]*z[0]+z[0]*S[0,1]*z[1]+z[1]*S[1,0]*z[0]+z[1]*S[1,1]*z[1])
```

1080 Ti yields
```
tensor(9.8596)
tensor(9.8596)
tensor(9.8596)
tensor(9.8596, device='cuda:0')
tensor(9.8596, device='cuda:0')
tensor(9.8596, device='cuda:0')
```

while the A40 yields
```
tensor(9.8596)
tensor(9.8596)
tensor(9.8596)
tensor(9.8616, device='cuda:0')
tensor(9.8616, device='cuda:0')
tensor(9.8596, device='cuda:0')
```

As seen, lines 4 and 5 of the output are incorrect when running the code on A40. Is this just a floating point issue or is it an actual bug?

### Versions

```
PyTorch version: 1.10.1+cu111
Is debug build: False
CUDA used to build PyTorch: 11.1
ROCM used to build PyTorch: N/A

OS: CentOS Linux 7 (Core) (x86_64)
GCC version: (GCC) 4.8.5 20150623 (Red Hat 4.8.5-44)
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.17

Python version: 3.9.7 (default, Sep 16 2021, 13:09:58)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-3.10.0-1160.42.2.el7.x86_64-x86_64-with-glibc2.17
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: NVIDIA A40
Nvidia driver version: 470.57.02
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.22.3
[pip3] torch==1.10.1+cu111
[pip3] torch-interpolations==0.1
[pip3] torchdiffeq==0.2.2
[conda] numpy                     1.22.3                   pypi_0    pypi
[conda] torch                     1.10.1+cu111             pypi_0    pypi
[conda] torch-interpolations      0.1                      pypi_0    pypi
[conda] torchdiffeq               0.2.2                    pypi_0    pypi
```

**Comment by ngimel** [comment]
It is a tf32 issue, disable it by setting `torch.backends.cuda.matmul.allow_tf32=False`. In the next pytorch version (1.12) and in the nightlies tf32 is disabled by default.

**Comment by xsacha** [comment]
Just a note, are these numerical errors generally fine for inference? What about training? Is it worse than FP16 for accuracy?

**Comment by ngimel** [comment]
Nvidia has run extensive experiments showing that networks trained with tf32 generally have the same convergence as those using fp32, same for inference. But when tf32 is used in isolated examples like above, or when matrix multiplications are used in image data preprocessing, the results can be surprising/less accurate.
